File: packages/duckdb/src/load/bundle.js
import fs from 'node:fs/promises';
import path from 'node:path';
import { cacheKey } from '../Cache.js';

async function retrieve(db, cache, sql, type) {
  const key = cacheKey(sql, type);
  const cached = cache.get(key);
  if (cached) return cached;
  switch (type) {
    case 'arrow':
      return db.arrowBuffer(sql);
    case 'json':
      return JSON.stringify(await db.query(sql));
    default:
      throw new Error(`Unsupported query type: ${type}`);
  }
}

export async function createBundle(db, cache, queries, dir) {
  const describe_re = /^DESCRIBE /;
  const pragma_re = /^PRAGMA /;
  const view_re = /^CREATE( TEMP| TEMPORARY)? VIEW/;
  const table_re = /^CREATE( TEMP| TEMPORARY)? TABLE( IF NOT EXISTS)? ([^\s]+)/;

  const manifest = { tables: [], queries: [] };

  await fs.mkdir(dir, { recursive: true });

  const querySet = new Set(queries);
  for (const query of querySet) {
    const sql = typeof query === 'string' ? query : query.sql;
    if (query.alias) {
      const table = query.alias;
      const file = path.resolve(dir, `${table}.parquet`);
      await db.exec(`COPY (${sql}) TO '${file}' (FORMAT PARQUET)`);
      manifest.tables.push(table);
    } else if (sql.startsWith('CREATE ')) {
      // table or view
      if (view_re.test(sql)) continue; // ignore views
      const table = sql.match(table_re)?.[3];
      const file = path.resolve(dir, `${table}.parquet`);
      await db.exec(`${sql}`);
      await db.exec(`COPY ${table} TO '${file}' (FORMAT PARQUET)`);
      manifest.tables.push(table);
    } else if (!pragma_re.test(sql)) {
      // select query
      const type = describe_re.test(sql) ? 'json' : 'arrow';
      const key = cacheKey(sql, type);
      const result = await retrieve(db, cache, sql, type);
      await fs.writeFile(path.resolve(dir, key), result);
      manifest.queries.push(key);
    }
  }

  await fs.writeFile(path.resolve(dir, 'bundle.json'), JSON.stringify(manifest, 0, 2));
  return manifest;
}

export async function loadBundle(db, cache, dir) {
  const manifest = JSON.parse(await fs.readFile(path.resolve(dir, 'bundle.json')));

  // load precomputed query results into the cache
  for (const key of manifest.queries) {
    const file = path.resolve(dir, key);
    const json = path.extname(file) === '.json';
    const data = await fs.readFile(file);
    cache.set(key, json ? JSON.parse(data) : data);
  }

  // load precomputed temp tables into the database
  for (const table of manifest.tables) {
    const file = path.resolve(dir, `${table}.parquet`);
    await db.exec(`CREATE TEMP TABLE IF NOT EXISTS ${table} AS SELECT * FROM '${file}'`);
  }
}

File: packages/duckdb-server/pkg/bundle.py
import os
import re
from functools import partial

import ujson

from pkg.query import get_arrow_bytes, get_json, get_key, retrieve


def create_bundle(con, cache, queries, directory):
    describe_re = re.compile(r"^DESCRIBE ")
    pragma_re = re.compile(r"^PRAGMA ")
    view_re = re.compile(r"^CREATE( TEMP| TEMPORARY)? VIEW")
    table_re = re.compile(r"^CREATE( TEMP| TEMPORARY)? TABLE( IF NOT EXISTS)? ([^\s]+)")

    manifest = {"tables": [], "queries": []}

    directory.mkdir(parents=True, exist_ok=True)

    for query in queries:
        sql = query if isinstance(query, str) else query.get("sql")

        if isinstance(query, dict) and query.get("alias"):
            table = query.get("alias")
            file = directory / f"{table}.parquet"
            con.execute(f"COPY ({sql}) TO '{file}' (FORMAT PARQUET)")
            manifest["tables"].append(table)

        elif sql.startswith("CREATE "):
            if view_re.match(sql):
                continue  # Ignore views

            table_match = table_re.match(sql)
            if table_match:
                table = table_match.group(3)
                file = directory / f"{table}.parquet"
                con.execute(sql)
                con.execute(f"COPY {table} TO '{file}' (FORMAT PARQUET)")
                manifest["tables"].append(table)

        elif not pragma_re.match(sql):
            command = "json" if describe_re.match(sql) else "arrow"
            key = get_key(sql, command)
            if command == "arrow":
                get = get_arrow_bytes
            elif command == "json":
                get = get_json
            else:
                raise ValueError(f"Unknown command {command}")
            result = retrieve(cache, {"sql": sql, "type": sql}, partial(get, con))
            with open(directory / key, "wb") as f:
                f.write(result)
            manifest["queries"].append(key)

    with open(directory / "bundle.json", "w") as f:
        ujson.dump(manifest, f, indent=2)

    return manifest


def load_bundle(con, cache, directory):
    with open(directory / "bundle.json") as f:
        manifest = ujson.load(f)

    # Load precomputed query results into the cache
    for key in manifest["queries"]:
        file = directory / key
        json_file = os.path.splitext(file)[1] == ".json"
        with open(file, "rb") as f:
            data = f.read()
            cache[key] = ujson.loads(data) if json_file else data

    # Load precomputed temp tables into the database
    for table in manifest["tables"]:
        file = directory / f"{table}.parquet"
        con.execute(f"CREATE TEMP TABLE IF NOT EXISTS {table} AS SELECT * FROM '{file}'")
